### PR TITLE
Added new datetime functions and aliases to PPL [Part 2]

### PR DESCRIPTION
**Pull request by matthewryanwells**
### Description
This PR adds multiple datetime functions that are in SQL to PPL.

The following functions are added:
- get_format
- last_day
- sec_to_time
- yearweek
- weekday
- to_seconds
- str_to_date
- timestampadd
- timestampdiff
- extract

### Issues Resolved
No related issue
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass, including unit test, integration test and doctest
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
  - [x] New functionality has user manual doc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by Yury-Fridlyand — APPROVED** [review]
<3

File: docs/user/ppl/functions/datetime.rst [hunk 1]
```diff
@@ -813,6 +813,78 @@ Example::
...
+EXTRACT
+-------
+
+Description
+>>>>>>>>>>>
+
+Usage: extract(part FROM date) returns a LONG with digits in order according to the given 'part' arguments.
+The specific format of the returned long is determined by the table below.
+
+Argument type: PART
```
Comment (GumpacG) [inline docs/user/ppl/functions/datetime.rst:825]: Missing new line I think
Screenshot 2023-07-06 at 5 13 41 PM

File: docs/user/ppl/functions/datetime.rst [hunk 2]
```diff
@@ -872,6 +944,29 @@ Examples::
...
+GET_FORMAT
+----------
+
+Description
+>>>>>>>>>>>
+
+Usage: Returns a string value containing string format specifiers based on the input arguments.
+
+Argument type: TYPE, STRING
+TYPE must be one of the following tokens: [DATE, TIME, DATETIME, TIMESTAMP].
```
Comment (GumpacG) [inline docs/user/ppl/functions/datetime.rst:956]: Same here missing new lines. You can view rich diff in the PR.

**Review by GumpacG — COMMENTED** [review]
Check all descriptions of Argument Type in the RST files have new lines.

**Comment by Yury-Fridlyand** [comment]
Hint: try RST validation tools - online or offline (e.g. VS code plugin).

File: docs/user/ppl/functions/datetime.rst (hunk 1 above)
Comment (matthewryanwells) [reply docs/user/ppl/functions/datetime.rst]: Good catch, I updated the documentation in https://github.com/Bit-Quill/opensearch-project-sql/pull/283/commits/872ece9c46db0eab82c2f34ed3cf7d973947c02a

File: docs/user/ppl/functions/datetime.rst (hunk 2 above)
Comment (matthewryanwells) [reply docs/user/ppl/functions/datetime.rst]: Updated in https://github.com/Bit-Quill/opensearch-project-sql/pull/283/commits/872ece9c46db0eab82c2f34ed3cf7d973947c02a

File: ppl/src/main/antlr/OpenSearchPPLParser.g4 [hunk 3]
```diff
@@ -534,10 +534,13 @@ dateTimeFunctionName
     | DAY_OF_MONTH
     | DAY_OF_WEEK
     | DAY_OF_YEAR
+    | EXTRACT
```
Comment (Yury-Fridlyand) [inline ppl/src/main/antlr/OpenSearchPPLParser.g4:537]: Please, copy to PPL full implementation of `extract` from #221. And for `get_format` function.

File: integ-test/src/test/java/org/opensearch/sql/ppl/DateTimeFunctionIT.java [hunk 4]
```diff
@@ -1062,4 +1062,85 @@ public void testTimeDiff() throws IOException {
...
+  @Test
+  public void testTimestampDiff() throws IOException{
+    var result = executeQuery(String.format("source=%s | eval f = timestampdiff('YEAR', '1997-01-01 00:00:00', '2001-03-06 00:00:00') | fields f", TEST_INDEX_DATE));
+    verifySchema(result, schema("f", null, "datetime"));
+    verifySome(result.getJSONArray("datarows"), rows(4));
+  }
+
+  @Test
+  public void testExtract() throws IOException{
+    var result = executeQuery(String.format("source=%s | eval f1 = extract('YEAR', '1997-01-01 00:00:00') | eval f2 = extract('MINUTE', time('10:17:36')) | fields f1, f2", TEST_INDEX_DATE));
```
Comment (Yury-Fridlyand) [inline integ-test/src/test/java/org/opensearch/sql/ppl/DateTimeFunctionIT.java:1138]: Please, update IT

File: integ-test/src/test/java/org/opensearch/sql/ppl/DateTimeFunctionIT.java [hunk 5]
```diff
@@ -1062,4 +1062,85 @@ public void testTimeDiff() throws IOException {
...
+    verifySchema(result,
+            schema("f1", null, "long"),
+            schema("f2", null, "long"));
+    verifySome(result.getJSONArray("datarows"), rows(1997L, 17L));
+  }
+
```
Comment (Yury-Fridlyand) [inline integ-test/src/test/java/org/opensearch/sql/ppl/DateTimeFunctionIT.java:1144]: Add ITs for `TIMESTAMPADD` and `TIMESTAMPDIFF`

File: ppl/src/test/java/org/opensearch/sql/ppl/antlr/PPLSyntaxParserTest.java [hunk 6]
```diff
@@ -7,11 +7,17 @@
...
+import static org.junit.Assert.assertThrows;
 
 import org.antlr.v4.runtime.tree.ParseTree;
 import org.junit.Rule;
 import org.junit.Test;
 import org.junit.rules.ExpectedException;
+import org.opensearch.sql.common.antlr.SyntaxCheckException;
+
+import java.util.List;
+import java.util.stream.Stream;
```
Comment (Yury-Fridlyand) [inline ppl/src/test/java/org/opensearch/sql/ppl/antlr/PPLSyntaxParserTest.java:20]: Reorder imports in sake of checkstyle

File: ppl/src/main/antlr/OpenSearchPPLParser.g4 (hunk 3 above)
Comment (matthewryanwells) [reply ppl/src/main/antlr/OpenSearchPPLParser.g4]: Thank you for your help with this, updated in https://github.com/Bit-Quill/opensearch-project-sql/pull/283/commits/1c64d5c6400e6de10759d9934ad0117e1c8123a4

File: integ-test/src/test/java/org/opensearch/sql/ppl/DateTimeFunctionIT.java (hunk 4 above)
Comment (matthewryanwells) [reply integ-test/src/test/java/org/opensearch/sql/ppl/DateTimeFunctionIT.java]: Updated in https://github.com/Bit-Quill/opensearch-project-sql/pull/283/commits/1c64d5c6400e6de10759d9934ad0117e1c8123a4

File: ppl/src/test/java/org/opensearch/sql/ppl/antlr/PPLSyntaxParserTest.java (hunk 6 above)
Comment (matthewryanwells) [reply ppl/src/test/java/org/opensearch/sql/ppl/antlr/PPLSyntaxParserTest.java]: Fixed all checkstyle failures in https://github.com/Bit-Quill/opensearch-project-sql/pull/283/commits/1c64d5c6400e6de10759d9934ad0117e1c8123a4

**Comment by Yury-Fridlyand** [comment]
You missed co copy a UT from SQL to PPL, jacoco fails on PPL

**Comment by matthewryanwells** [comment]
> You missed co copy a UT from SQL to PPL, jacoco fails on PPL

Just noticed that and pushed a fix

**Review by Yury-Fridlyand — APPROVED** [review]

File: docs/user/ppl/functions/datetime.rst [hunk 7]
```diff
@@ -1637,6 +1802,55 @@ Example::
...
+
+Description
+>>>>>>>>>>>
+
+Usage: Returns a DATETIME value based on a passed in DATE/DATETIME/TIME/TIMESTAMP/STRING argument and an INTERVAL and INTEGER argument which determine the amount of time to be added.
+If the third argument is a STRING, it must be formatted as a valid DATETIME. If only a TIME is provided, a DATETIME is still returned with the DATE portion filled in using the current date.
+If the third argument is a DATE, it will be automatically converted to a DATETIME.
+
+Argument type: INTERVAL, INTEGER, DATE/DATETIME/TIME/TIMESTAMP/STRING
+INTERVAL must be one of the following tokens: [MICROSECOND, SECOND, MINUTE, HOUR, DAY, WEEK, MONTH, QUARTER, YEAR]
```
Comment (GumpacG) [inline docs/user/ppl/functions/datetime.rst:1816]: This also needs a new line.

File: docs/user/ppl/functions/datetime.rst [hunk 8]
```diff
@@ -1637,6 +1802,55 @@ Example::
...
+Description
+>>>>>>>>>>>
+
+Usage: TIMESTAMPDIFF(interval, start, end) returns the difference between the start and end date/times in interval units.
+If a TIME is provided as an argument, it will be converted to a DATETIME with the DATE portion filled in using the current date.
+Arguments will be automatically converted to a DATETIME/TIME/TIMESTAMP when appropriate.
+Any argument that is a STRING must be formatted as a valid DATETIME.
+
+Argument type: INTERVAL, DATE/DATETIME/TIME/TIMESTAMP/STRING, DATE/DATETIME/TIME/TIMESTAMP/STRING
+INTERVAL must be one of the following tokens: [MICROSECOND, SECOND, MINUTE, HOUR, DAY, WEEK, MONTH, QUARTER, YEAR]
```
Comment (GumpacG) [inline docs/user/ppl/functions/datetime.rst:1841]: This also needs a new line.

File: ppl/src/test/java/org/opensearch/sql/ppl/antlr/PPLSyntaxParserTest.java [hunk 9]
```diff
@@ -268,5 +272,56 @@ public void testDescribeCommandWithSourceShouldFail() {
 
     new PPLSyntaxParser().parse("describe source=t");
   }
+
+  @Test
+  public void testCan_parse_extract_function() {
```
Comment (GumpacG) [inline ppl/src/test/java/org/opensearch/sql/ppl/antlr/PPLSyntaxParserTest.java:277]: Unfortunately there's a mix of camel case and snake case between different files. However, we should try to make it consistent within a file. I think the names should be camel case based on already existing test names.

**Review by GumpacG — CHANGES_REQUESTED** [review]

File: docs/user/ppl/functions/datetime.rst (hunk 8 above)
Comment (matthewryanwells) [reply docs/user/ppl/functions/datetime.rst:1843]: added in https://github.com/Bit-Quill/opensearch-project-sql/pull/283/commits/90424e828d0fde750546bfb25621f78e4c3135d0

**Review by GumpacG — APPROVED** [review]

**Review by Yury-Fridlyand — APPROVED** [review]
CI fails due to `log4j` update on upstream. We can ignore that for now.

**Comment by Yury-Fridlyand** [comment]
@matthewryanwells you need one more rebase now to get CI working.

**Review by Yury-Fridlyand — APPROVED** [review]

**Pull request merged**